datafusion/datasource/src/url.rs

```diff
@@ -324,33 +324,105 @@ impl ListingTableUrl {
     }
 }
 
+/// Lists files with cache support, using prefix-aware lookups.
+///
+/// # Arguments
+/// * `ctx` - The session context
+/// * `store` - The object store to list from
+/// * `full_prefix` - The full prefix to list (table_base + partition prefix)
+/// * `table_base_path` - The table's base path (the stable cache key)
+///
+/// # Cache Behavior :
+/// The cache key is always `table_base_path`. When a partition-specific listing
+/// is requested (full_prefix includes partition path), the cache:
+/// - Looks up `table_base_path` in the cache
+/// - Filters results to match `full_prefix`
+/// - Returns filtered results without a storage call
+///
+/// On cache miss, the full table is always listed and cached, ensuring
+/// subsequent partition queries can be served from cache.
 async fn list_with_cache<'b>(
     ctx: &'b dyn Session,
     store: &'b dyn ObjectStore,
-    prefix: &Path,
+    full_prefix: &Path,
+    table_base_path: &Path,
 ) -> Result<BoxStream<'b, Result<ObjectMeta>>> {
     match ctx.runtime_env().cache_manager.get_list_files_cache() {
         None => Ok(store
-            .list(Some(prefix))
+            .list(Some(full_prefix))
             .map(|res| res.map_err(|e| DataFusionError::ObjectStore(Box::new(e))))
             .boxed()),
         Some(cache) => {
-            let vec = if let Some(res) = cache.get(prefix) {
-                debug!("Hit list all files cache");
+            // Compute the relative prefix (partition path relative to table base)
+            let relative_prefix = compute_relative_prefix(table_base_path, full_prefix);
```

Comment:
I'm feeling a bit confused with the structure of this code. The calling method already has the knowledge of the table's base path vs a user's provided prefix, and it concatenates them into a full listing path here:
https://github.com/Yuvraj-cyborg/datafusion/blob/47e1bcf9cb91b83fe0b46d7bc0cfa232fa60e2cd/datafusion/datasource/src/url.rs#L248-L254

Rather than performing that concatenation and then deconstructing the path, can we just pass in `table_base_path` and `prefix`?